src/app/hooks/useAPI.js

```diff
@@ -0,0 +1,91 @@
+import { useState, useEffect, useRef, useReducer } from 'react';
+import PropTypes from 'prop-types';
+
+import api from '~config/api';
+
+const dataFetchReducer = (state, action) => {
+  switch (action.type) {
+    case 'FETCH_INIT':
+      return {
+        ...state,
+        isLoading: true,
+        isError: ''
+      };
+    case 'FETCH_SUCCESS':
+      return {
+        ...state,
+        response: action.payload,
+        isLoading: false,
+        isError: ''
+      };
+    case 'FETCH_FAILURE':
+      return {
+        ...state,
+        response: action.payload,
+        isLoading: false,
+        isError: 'Something Wrong Happened'
+      };
+    default:
+      return { ...state };
+  }
+};
+
+const useAPI = (config, initialData = {}, execute = false) => {
+  const [state, dispatch] = useReducer(dataFetchReducer, {
+    isLoading: false,
+    isError: '',
+    response: initialData
+  });
+  const [conf, setConf] = useState(config);
```

Comment:
why store a config in the state,  this could cause problems of old state

Reply:
the idea of using config in a state, serves so that at the time of adding or modifying that state, useEffect is activated to make the desired request.

`setConf(prevState => ({
      ...prevState,
      ...params
    }));`

the status is updated, updating or adding desired fields